### PR TITLE
Implement Serializable for all consensus wire messages

### LINKED ISSUE #22: Make serialization of consensus messages fully compatible with dusk wire protocol
#### Summary
Current serialization of consensus messages (NewBlock, Reduction, Agreement) does not match fully the serialization of the dusk wire protocol. This issue should also address https://github.com/dusk-network/consensus/pull/21#discussion_r1006891657

#### Additional context
A discussion should be open about if we should update the dusk wire protocol having in mind that it is not fully efficient (e.g fixed-size fields like signature are serialized like variable-size fields)

### PR DESCRIPTION
**Pull request by goshawk-3**
fixes #22

### CODE REVIEW

File: src/messages.rs [hunk 1]
```diff
@@ -4,25 +4,49 @@
...
+        w.write_all(buf)?;
+
+        Ok(())
+    }
+
+    fn read_var_le_bytes<R: Read, const N: usize>(
+        r: &mut R,
+    ) -> io::Result<[u8; N]> {
+        let mut buf = [0u8; 1];
+        r.read_exact(&mut buf)?;
```
Comment (herr-seppia) [inline src/messages.rs:43]: Why are you reading the first byte (I believe is `len`) and then discard it?

File: src/messages.rs [hunk 2]
```diff
@@ -43,57 +67,55 @@ pub struct Message {
...
-
-        let mut buf = BytesMut::with_capacity(payload_as_vec.len());
-        buf.put(&self.header.to_bytes()[..]);
-        buf.put(&payload_as_vec[..]);
-        buf.to_vec()
+    fn write<W: Write>(&self, w: &mut W) -> io::Result<()> {
+        self.header.write(w)?;
+
+        match &self.payload {
+            Payload::NewBlock(p) => p.write(w),
```
Comment (herr-seppia) [inline src/messages.rs:81]: Are these `write` returning a `Result`? if so you should add the `?`

File: src/messages.rs [hunk 3]
```diff
@@ -179,34 +201,61 @@ pub struct Header {
...
+    where
+        Self: Sized,
+    {
+        // Read topic
+        let mut buf = [0u8; 1];
+        r.read_exact(&mut buf)?;
+        let topic = buf[0];
+
+        // Read bls pubkey
+        let buf: [u8; 96] = Self::read_var_le_bytes(r)?;
```
Comment (herr-seppia) [inline src/messages.rs:225]: Do we need a var_le_bytes even if we know the exact size of the field?

File: src/messages.rs [hunk 4]
```diff
@@ -276,12 +325,12 @@ impl Header {
...
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub enum Payload {
     Reduction(payload::Reduction),
     NewBlock(Box<payload::NewBlock>),
     StepVotes(payload::StepVotes),
-    StepVotesWithCandidate(payload::StepVotesWithCandidate),
+    StepVotesWithCandidate(Box<payload::StepVotesWithCandidate>),
```
Comment (herr-seppia) [inline src/messages.rs:333]: Why did you added the `Box`?

File: src/messages.rs [hunk 5]
```diff
@@ -296,28 +345,26 @@ impl Default for Payload {
...
         }
 
-        fn from_bytes(buf: &mut Bytes) -> Self {
-            let mut r = Self {
-                signed_hash: [0; 48],
-            };
-
-            buf.copy_to_slice(&mut r.signed_hash);
-            r
+        /// Deserialize struct from buf by consuming N bytes.
```
Comment (herr-seppia) [inline src/messages.rs:361]: Is this comment a leftover?

File: src/messages.rs [hunk 6]
```diff
@@ -347,26 +394,26 @@ pub mod payload {
...
-            buf.put(&self.signed_hash[..]);
-            buf.put(&candidate_as_bytes[..]);
-
-            buf.to_vec()
+            Ok(())
         }
 
-        fn from_bytes(buf: &mut Bytes) -> Self {
-            let mut nb = NewBlock::default();
+        /// Deserialize struct from buf by consuming N bytes.
```
Comment (herr-seppia) [inline src/messages.rs:405]: Ditto

File: src/selection/block_generator.rs [hunk 7]
```diff
@@ -87,9 +87,10 @@ impl<T: Operations> Generator<T> {
             generator_bls_pubkey: *pubkey.bytes(),
             state_hash: [0; 32],
             hash: [0; 32],
+            cert: Certificate::default(),
         };
 
-        Ok(Block::new(blk_header, vec![]))
+        Ok(Block::new(blk_header, vec![]).expect("block should be valid"))
```
Comment (herr-seppia) [inline src/selection/block_generator.rs:93]: Should we reprobavate the error instead of panicking?

**Review by herr-seppia — CHANGES_REQUESTED** [review]
I like it

However there are some parts that are still not clear to me, so I'm not sure to approve it yet

File: src/messages.rs (hunk 4 above)
Comment (goshawk-3) [reply src/messages.rs:333]: Due to a clippy warning for having too big variant.

File: src/messages.rs (hunk 2 above)
Comment (goshawk-3) [reply src/messages.rs:81]: No need for `?` here.  I return `write` result instead.

File: src/messages.rs (hunk 3 above)
Comment (goshawk-3) [reply src/messages.rs:225]: Unfortunately yes, we need it. This is again to be compatible with wire protocol (golang) implementation.

File: src/selection/block_generator.rs (hunk 7 above)
Comment (goshawk-3) [reply src/selection/block_generator.rs:93]: While testing is progress, I'd like to have fail-fast approach.

File: src/messages.rs (hunk 5 above)
Comment (goshawk-3) [reply src/messages.rs]: Removed

File: src/messages.rs (hunk 6 above)
Comment (goshawk-3) [reply src/messages.rs]: Removed

**Review by herr-seppia — APPROVED** [review]
LGTM

**Pull request merged**